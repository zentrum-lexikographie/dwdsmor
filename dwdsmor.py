#!/usr/bin/python3
# dwdsmor.py - analyse word forms with DWDSmor
# Gregor Middell and Andreas Nolda 2022-08-18

import sys
import os
import argparse
import re
import sfst_transduce
import csv
import json
from blessings import Terminal
from collections import namedtuple
from functools import cached_property

version = 5.1

basedir = os.path.dirname(__file__)
libdir  = os.path.join(basedir, "lib")
libfile = os.path.join(libdir, "smor-full.ca")

Component = namedtuple("Component", ["form", "lemma", "tags"])

class Analysis(tuple):
    def __new__(cls, analysis, components):
        inst = tuple.__new__(cls, components)
        inst.analysis = analysis
        return inst

    @cached_property
    def form(self):
        return "".join(a.form for a in self)

    @cached_property
    def lemma(self):
        return "".join(a.lemma for a in self)

    @cached_property
    def segmented_lemma(self):
        return re.sub(r"(<IDX[^>]+>)?(<\^ABBR>)?<\+[^>]+>.*", "", self.analysis)

    @cached_property
    def tags(self):
        return [tag for a in self for tag in a.tags]

    @cached_property
    def index(self):
        for tag in self.tags:
            if tag.startswith("IDX"):
                return tag[3:]

    @cached_property
    def pos(self):
        for tag in self.tags:
            if tag.startswith("+"):
                return tag[1:]

    _subcat_tags       = {"Pers": True, "Refl": True, "Def": True, "Indef": True, "Neg": True}
    _auxiliary_tags    = {"haben": True, "sein": True}
    _degree_tags       = {"Pos": True, "Comp": True, "Sup": True}
    _person_tags       = {"1": True, "2": True, "3": True}
    _gender_tags       = {"Fem": True, "Neut": True, "Masc": True, "NoGend": True, "Invar": True}
    _case_tags         = {"Nom": True, "Gen": True, "Dat": True, "Acc": True, "Invar": True}
    _number_tags       = {"Sg": True, "Pl": True, "Invar": True}
    _inflection_tags   = {"St": True, "Wk": True, "NoInfl": True, "Invar": True}
    _function_tags     = {"Attr": True, "Subst": True, "Pred": True, "Adv": True}
    _nonfinite_tags    = {"Inf": True, "PPres": True, "PPast": True}
    _mood_tags         = {"Ind": True, "Subj": True, "Imp": True}
    _tense_tags        = {"Pres": True, "Past": True}
    _abbreviation_tags = {"^ABBR": True}
    _metainfo_tags     = {"Old": True, "NonSt": True, "CAP": True}

    def tag_of_type(self, type_map):
        for tag in self.tags:
            if tag in type_map:
                return tag

    @cached_property
    def subcat(self):
        return self.tag_of_type(Analysis._subcat_tags)

    @cached_property
    def auxiliary(self):
        return self.tag_of_type(Analysis._auxiliary_tags)

    @cached_property
    def degree(self):
        return self.tag_of_type(Analysis._degree_tags)

    @cached_property
    def person(self):
        return self.tag_of_type(Analysis._person_tags)

    @cached_property
    def gender(self):
        return self.tag_of_type(Analysis._gender_tags)

    @cached_property
    def case(self):
        return self.tag_of_type(Analysis._case_tags)

    @cached_property
    def number(self):
        return self.tag_of_type(Analysis._number_tags)

    @cached_property
    def inflection(self):
        return self.tag_of_type(Analysis._inflection_tags)

    @cached_property
    def function(self):
        return self.tag_of_type(Analysis._function_tags)

    @cached_property
    def nonfinite(self):
        return self.tag_of_type(Analysis._nonfinite_tags)

    @cached_property
    def mood(self):
        return self.tag_of_type(Analysis._mood_tags)

    @cached_property
    def tense(self):
        return self.tag_of_type(Analysis._tense_tags)

    @cached_property
    def abbreviation(self):
        if self.tag_of_type(Analysis._abbreviation_tags):
            return "yes"

    @cached_property
    def metainfo(self):
        return self.tag_of_type(Analysis._metainfo_tags)

    def as_dict(self):
        return {"form": self.form,
                "analysis": self.analysis,
                "lemma": self.lemma,
                "segmentedlemma": self.segmented_lemma,
                "index": self.index,
                "pos": self.pos,
                "subcat": self.subcat,
                "auxiliary": self.auxiliary,
                "degree": self.degree,
                "person": self.person,
                "gender": self.gender,
                "case": self.case,
                "number": self.number,
                "inflection": self.inflection,
                "function": self.function,
                "nonfinite": self.nonfinite,
                "mood": self.mood,
                "tense": self.tense,
                "abbreviation": self.abbreviation,
                "metainfo": self.metainfo}

    _empty_component_texts = set(["", ":"])
    _curly_braces_re = re.compile(r"[{}]")

    def _decode_component_text(text):
        lemma = ""
        form  = ""
        text_len = len(text)
        ti = 0
        prev = None
        while ti < text_len:
            current = text[ti]
            nti = ti + 1
            next = text[nti] if nti < text_len else None
            if current == ":":
                lemma += prev or ""
                form  += next or ""
                ti += 1
            elif next != ":":
                lemma += current
                form  += current
            ti += 1
            prev = current
        return {"lemma": lemma, "form": form}

    def _decode_analysis(analysis):
        # "QR-Code" -> "{:<>QR}:<>-<TRUNC>:<>Code<+NN>:<><Masc>:<><Acc>:<><Sg>:<>"
        analysis = Analysis._curly_braces_re.sub("", analysis)
        for a in re.finditer(r"([^<]*)(?:<([^>]*)>)?", analysis):
            text = a.group(1)
            tag  = a.group(2) or ""
            component = Analysis._decode_component_text(text)
            if tag != "":
                component["tag"] = tag
            yield component

    def _join_tags(components):
        result = []
        current = None
        for c in components:
            c = c.copy()
            if current is None or c["form"] != "" or c["lemma"] != "":
                c["tags"] = []
                result.append(c)
                current = c
            if "tag" in c:
                current["tags"].append(c["tag"])
                del c["tag"]
        return result

    def _join_untagged(components):
        result = []
        buf = []
        for c in components:
            buf.append(c)
            if len(c["tags"]) > 0:
                joined = {"lemma": "", "form": "", "tags": []}
                for c in buf:
                    joined["lemma"] += c["lemma"]
                    joined["form"]  += c["form"]
                    joined["tags"]  += c["tags"]
                result.append(joined)
                buf = []
        if len(buf) > 0:
            result = result + buf
        return result

def parse(analyses):
    component_list = []
    for analysis in analyses:
        components = Analysis._decode_analysis(analysis)
        components = Analysis._join_tags(components)
        components = Analysis._join_untagged(components)
        if not components in component_list:
            component_list.append(components)
            yield Analysis(analysis, [Component(**c) for c in components])

def analyse_word(transducer, word):
    return parse(transducer.analyse(word))

def analyse_words(transducer, words):
    return tuple(analyse_word(transducer, word) for word in words)

def string(value):
    return str(value or "")

def output_json(words, analyses, output):
    word_analyses = []
    for word, analysis in zip(words, analyses):
        word_analyses.append({"word": word,
                              "analyses": [a.as_dict() for a in analysis]})
    json.dump(word_analyses, output, ensure_ascii=False)

def output_dsv(words, analyses, output, header=True, force_color=False, delimiter="\t"):
    term = Terminal(force_styling=force_color)
    csv_writer = csv.writer(output, delimiter=delimiter)
    if header:
        csv_writer.writerow([term.bold("Wordform"),
                             term.bright_black("Analysis"),
                             term.bold_underline("Lemma"),
                             term.bright_black_underline("Segmentation"),
                             term.underline("Index"),
                             term.underline("POS"),
                             term.underline("Subcategory"),
                             term.underline("Auxiliary"),
                             "Degree",
                             "Person",
                             "Gender",
                             "Case",
                             "Number",
                             "Inflection",
                             "Function",
                             "Nonfinite",
                             "Mood",
                             "Tense",
                             "Abbreviation",
                             "Metainfo"])
    for word, analysis in zip(words, analyses):
        for a in analysis:
            csv_writer.writerow([term.bold(word),
                                 term.bright_black(a.analysis),
                                 term.bold_underline(a.lemma),
                                 term.bright_black_underline(a.segmented_lemma),
                                 term.underline(string(a.index)),
                                 term.underline(string(a.pos)),
                                 term.underline(string(a.subcat)),
                                 term.underline(string(a.auxiliary)),
                                 a.degree,
                                 a.person,
                                 a.gender,
                                 a.case,
                                 a.number,
                                 a.inflection,
                                 a.function,
                                 a.nonfinite,
                                 a.mood,
                                 a.tense,
                                 a.abbreviation,
                                 a.metainfo])

def output_analyses(transducer, input, output, header=True, force_color=False, output_format="tsv"):
    words = tuple(word.strip() for word in input.readlines() if word.strip())
    analyses = analyse_words(transducer, words)
    if analyses:
        if output_format == "json":
            output_json(words, analyses, output)
        elif output_format == "csv":
            output_dsv(words, analyses, output, header, force_color, delimiter=",")
        else:
            output_dsv(words, analyses, output, header, force_color)

def main():
    e = False
    try:
        parser = argparse.ArgumentParser()
        parser.add_argument("input", nargs="?", type=argparse.FileType("r"), default=sys.stdin,
                            help="input file (one word form per line; default: stdin)")
        parser.add_argument("output", nargs="?", type=argparse.FileType("w"), default=sys.stdout,
                            help="output file (default: stdout)")
        parser.add_argument("-c", "--csv", action="store_true",
                            help="output CSV table")
        parser.add_argument("-C", "--force-color", action="store_true",
                            help="preserve color and formatting when piping output")
        parser.add_argument("-H", "--no-header", action="store_false",
                            help="suppress table header")
        parser.add_argument("-j", "--json", action="store_true",
                            help="output JSON object")
        parser.add_argument("-t", "--transducer", default=libfile,
                            help="path to transducer file (default: {0})".format(os.path.relpath(libfile, os.getcwd())))
        parser.add_argument("-v", "--version", action="version",
                            version="{0} {1}".format(parser.prog, version))
        args = parser.parse_args()
        term = Terminal(force_styling=args.force_color)
        transducer = sfst_transduce.CompactTransducer(args.transducer)
        transducer.both_layers = False
        if args.json:
            output_format = "json"
        elif args.csv:
            output_format = "csv"
        else:
            output_format = "tsv"
        output_analyses(transducer, args.input, args.output, args.no_header, args.force_color, output_format)
    except KeyboardInterrupt:
        sys.exit(130)
    # except TypeError:
    #     print(term.bold_red(args.transducer) + ": No such transducer file.", file=sys.stderr)
    #     e = True
    # except RuntimeError:
    #     print(term.bold_red(args.transducer) + ": No such transducer.", file=sys.stderr)
    #     e = True
    if e:
        exit = 1
    else:
        exit = 0
    return exit

if __name__ == "__main__":
    sys.exit(main())
