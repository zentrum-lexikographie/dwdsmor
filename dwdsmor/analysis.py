from collections import namedtuple
from functools import cached_property

import csv
import json
import re

import click
import jellyfish

import sfst

Morpheme = namedtuple('Morpheme', ['word', 'lemma', 'tags'])


class Analysis(tuple):
    @staticmethod
    def parse(analyses):
        for analysis in analyses:
            morphemes = Analysis._decode_analysis(analysis)
            morphemes = Analysis._join_tags(morphemes)
            morphemes = Analysis._join_untagged(morphemes)
            yield Analysis(analysis, [Morpheme(**m) for m in morphemes])

    def __new__(cls, analysis, morphemes):
        inst = tuple.__new__(cls, morphemes)
        inst.analysis = analysis
        return inst

    @cached_property
    def tags(self):
        return [tag for m in self for tag in m.tags]

    @cached_property
    def pos(self):
        for tag in self.tags:
            if tag.startswith('+'):
                return tag[1:]

    @cached_property
    def word(self):
        return ''.join(m.word for m in self)

    @cached_property
    def lemma(self):
        return ''.join(m.lemma for m in self)

    _genus_tags = {'Fem': True, 'Neut': True, 'Masc': True}
    _numerus_tags = {'Sg': True, 'Pl': True}
    _person_tags = {'1': True, '2': True, '3': True}
    _casus_tags = {'Nom': True, 'Gen': True, 'Dat': True, 'Acc': True}
    _tempus_tags = {'Pres': True, 'Past': True, 'PPast': True}

    def tag_of_type(self, type_map):
        for tag in self.tags:
            if tag in type_map:
                return tag

    @cached_property
    def genus(self):
        return self.tag_of_type(Analysis._genus_tags)

    @cached_property
    def numerus(self):
        return self.tag_of_type(Analysis._numerus_tags)

    @cached_property
    def person(self):
        return self.tag_of_type(Analysis._person_tags)

    @cached_property
    def casus(self):
        return self.tag_of_type(Analysis._casus_tags)

    @cached_property
    def tempus(self):
        return self.tag_of_type(Analysis._tempus_tags)

    @cached_property
    def dist_score(self):
        return jellyfish.levenshtein_distance(
            self.word.lower(), self.lemma.lower()
        )

    def as_dict(self):
        return {
            'word': self.word,
            'analysis': self.analysis,
            'lemma': self.lemma,
            'pos': self.pos,
            'genus': self.genus,
            'numerus': self.numerus,
            'casus': self.casus,
            'person': self.person,
            'tempus': self.tempus,
            'morphemes': [m._asdict() for m in self]
        }

    _empty_component_texts = set(['', ':'])
    _curly_braces_re = re.compile(r'[{}]')

    def _decode_component_text(text):
        lemma = ''
        word = ''
        text_len = len(text)
        ti = 0
        prev = None
        while ti < text_len:
            current = text[ti]
            nti = ti + 1
            next = text[nti] if nti < text_len else None
            if current == ':':
                lemma += prev or ''
                word += next or ''
                ti += 1
            elif next != ':':
                lemma += current
                word += current
            ti += 1
            prev = current
        return {'lemma': lemma, 'word': word}

    def _decode_analysis(analysis):
        # 'QR-Code' -> '{:<>QR}:<>-<TRUNC>:<>Code<+NN>:<><Masc>:<><Acc>:<><Sg>:<>'
        analysis = Analysis._curly_braces_re.sub('', analysis)

        for m in re.finditer(r'([^<]*)(?:<([^>]*)>)?', analysis):
            text = m.group(1)
            tag = m.group(2) or ''
            component = Analysis._decode_component_text(text)
            if tag != '':
                component['tag'] = tag
            yield component

    def _join_tags(analysis):
        result = []
        current = None
        for c in analysis:
            c = c.copy()
            if current is None or c['word'] != '' or c['lemma'] != '':
                c['tags'] = []
                result.append(c)
                current = c
            if 'tag' in c:
                current['tags'].append(c['tag'])
                del c['tag']
        return result

    def _join_untagged(analysis):
        result = []
        buf = []
        for c in analysis:
            buf.append(c)
            if len(c['tags']) > 0:
                joined = {'lemma': '', 'word': '', 'tags': []}
                for c in buf:
                    joined['lemma'] += c['lemma']
                    joined['word'] += c['word']
                    joined['tags'] += c['tags']
                result.append(joined)
                buf = []
        if len(buf) > 0:
            result = result + buf
        return result


def create_transducer(file_name):
    try:
        smor = sfst.CompactTransducer(file_name)
        smor.both_layers = True
        return smor
    except RuntimeError:
        raise click.ClickException(
            f'Cannot instantiate transducer from {file_name}'
        )


@click.command()
@click.option(
    '-a', '--automaton',
    required=True,
    type=click.Path(exists=True, dir_okay=False),
    help='Path to the FST (compacted, i.e. *.ca)'
)
@click.option(
    '--csv', 'output_format',
    flag_value='csv',
    default=True,
    help='Output in CSV format (default)'
)
@click.option(
    '--json', 'output_format',
    flag_value='json',
    help='Output in JSON format'
)
@click.argument(
    'input',
    type=click.File('r'),
    default='-'
)
@click.argument(
    'output',
    type=click.File('w'),
    default='-'
)
def main(automaton, output_format, input, output):
    '''DWDSmor Morphological Analysis

    Copyright (C) 2021 Berlin-Brandenburgische Akademie der Wissenschaften

    Feeds word forms – given as one form per line – into a finite state
    transducer and outputs the results – valid paths through the transducer
    representing possible analyses.

    INPUT is the file with word forms to analyze (defaults to stdin).

    OUTPUT is the file to write results to (defaults to stdout).'''
    smor = create_transducer(click.format_filename(automaton))
    words = tuple(map(lambda l: l.strip(), input.readlines()))
    analyses = tuple([Analysis.parse(smor.analyse(word)) for word in words])
    if output_format == 'json':
        json.dump({
            word: [a.as_dict() for a in analysis]
            for word, analysis in zip(words, analyses)
        }, output)
    elif output_format == 'csv':
        csv_writer = csv.writer(output)
        csv_writer.writerow([
            "Word", "Analysis", "Lemma", "POS",
            "Genus", "Numerus", "Casus", "Person", "Tempus"
        ])
        for word, analysis in zip(words, analyses):
            for a in analysis:
                csv_writer.writerow([
                    word, a.analysis, a.lemma, a.pos,
                    a.genus, a.numerus, a.casus, a.person, a.tempus
                ])


if __name__ == '__main__':
    main()
