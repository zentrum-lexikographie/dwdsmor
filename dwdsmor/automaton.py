"""
Handling of SFST automata/transducer collections on which DWDSmor is based.

For different use cases like lemmatization, paradigm generation or
morphological segmentation, DWDSmor relies on different transducers
based on a common lexicon. Those transducers are stored collectively
in a common local directory from which they can be read and used via
the class ``Automata`` in this module.
"""

__all__ = [
    "detect_dev_root_dir",
    "detect_root_dir",
    "Generator",
    "Analyzer",
    "Automata",
    "automata",
    "Lemmatizer",
    "lemmatizer",
    "load_from_hub",
]

import csv
import lzma
import os
from collections import defaultdict
from dataclasses import dataclass
from os import PathLike
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Union

from dotenv import dotenv_values
from huggingface_hub import snapshot_download
from huggingface_hub.utils import disable_progress_bars, enable_progress_bars
from sfst_transduce import CompactTransducer, Transducer

from .log import logger
from .traversal import Traversal
from .version import __version__

config = {
    **dotenv_values(".env.shared"),
    **dotenv_values(".env"),
    **os.environ,
}


class AutomataDirNotFound(Exception):
    pass


def detect_dev_root_dir():
    """
    Detect development environment and return one of its automaton
    dirs.

    Development environments are flagged via a set environment
    variable ``DWDSMOR_ENV`` (i. e. in ``.env.shared``). In such
    environments, one of the automata in a subdirectory of ``build/``,
    i.e. ``build/open`` is returned.
    """
    if config.get("DWDSMOR_DEV"):
        build_dir = Path("build")
        for edition in ("dwds", "open", "dev"):
            edition_dir = build_dir / edition
            if edition_dir.is_dir():
                return str(edition_dir)
        raise AutomataDirNotFound(
            "$DWDSMOR_DEV is set, but no known automata available under build/"
        )


def detect_root_dir() -> Path:
    """
    Detect local automata directory if none is specified.

    1. Try to get it from an environment variable ``DWDSMOR_AUTOMATA_DIR``.
    2. Try to detect a development environment and get it from there.
    3. Raise ``AutomataDirNotFound``.
    """
    root = config.get("DWDSMOR_AUTOMATA_DIR")
    if root:
        return Path(root)
    root = detect_dev_root_dir()
    if root:
        return Path(root)
    raise AutomataDirNotFound()


@dataclass
class Generator:
    """
    Wrap a SFST transducer for generating strings, i.e. inflection paradigms.
    """

    transducer: Transducer

    def generate(self, s: str) -> Iterable[Traversal]:
        """
        Generate all traversals of a transducer for the given input string.
        """
        return (Traversal.parse(traversal) for traversal in self.transducer.generate(s))


@dataclass
class Analyzer:
    """
    Wrap a SFST transducer providing a morphological analysis of input strings.
    """

    transducer: CompactTransducer

    def analyze(self, s: str) -> Iterable[Traversal]:
        """
        Generate all traversals of a transducer that express
        morphological analyses of an input string.
        """
        return (Traversal.parse(traversal) for traversal in self.transducer.analyse(s))


automaton_types = {"finite", "index", "lemma", "morph", "root"}
traversal_automaton_types = {"index"}


def assert_valid_automaton_type(automaton_type: str, types=automaton_types):
    assert automaton_type in automaton_types, (
        f"{automaton_type} is not a valid automaton_type. "
        f"Supported types: {automaton_types}"
    )


class Automata:
    def __init__(self, root_dir: Optional[Union[str, PathLike]] = None):
        self.root_dir = Path(root_dir) if root_dir else detect_root_dir()

    def transducer(
        self, automaton_type="lemma", generate=True, both_layers=False
    ) -> Union[Transducer, CompactTransducer]:
        file_suffix = "a" if generate else "ca"
        transducer_file = Path(self.root_dir) / f"{automaton_type}.{file_suffix}"
        assert transducer_file.is_file(), f"{transducer_file} does not exist"
        if not generate:
            transducer = CompactTransducer(transducer_file.as_posix())
            transducer.both_layers = both_layers
            return transducer
        else:
            return Transducer(transducer_file.as_posix())

    def generator(self, automaton_type="index") -> Generator:
        assert_valid_automaton_type(automaton_type)
        return Generator(self.transducer(automaton_type))  # type: ignore

    def analyzer(self, automaton_type="lemma", both_layers=False) -> Analyzer:
        assert_valid_automaton_type(automaton_type)
        return Analyzer(
            self.transducer(
                automaton_type, generate=False, both_layers=both_layers
            )  # type: ignore
        )

    def traversals(self, automaton_type="index") -> Dict[str, List[Traversal]]:
        assert_valid_automaton_type(automaton_type, types=traversal_automaton_types)
        traversals_file = Path(self.root_dir) / f"{automaton_type}.csv.lzma"
        traversals = defaultdict(list)
        with lzma.open(traversals_file, "rt") as traversals_csv:
            for traversal in csv.DictReader(traversals_csv):
                traversal_obj = Traversal(**traversal)
                traversals[traversal_obj.analysis].append(traversal_obj)
        return traversals


default_repo_id = config.get("DWDSMOR_HF_REPO_ID", "zentrum-lexikographie/dwdsmor-open")


def load_from_hub(
    repo_id: Optional[str] = None, revision: Optional[str] = None, **kwargs
):
    repo_id = repo_id or default_repo_id
    revision = revision or f"v{__version__}"
    logger.debug("Load automata from Huggingface repo %s @ %s", repo_id, revision)
    try:
        disable_progress_bars()
        return Automata(snapshot_download(repo_id=repo_id, revision=revision, **kwargs))
    finally:
        enable_progress_bars()


def automata(automata_location: Optional[str] = None, *args, **kwargs):
    if automata_location is not None:
        path = Path(automata_location)
        if path.is_dir():
            logger.debug("Load automata from local dir '%s'", str(path))
            return Automata(path)
    if automata_location is None:
        try:
            detected_dir = detect_root_dir()
            logger.debug(
                "Load automata from detected local dir '%s'", str(detected_dir)
            )
            return Automata(detected_dir)
        except AutomataDirNotFound:
            pass
    return load_from_hub(automata_location, *args, **kwargs)


class Lemmatizer:
    def __init__(self, automata, automaton_type="lemma"):
        self.analyzer = automata.analyzer(automaton_type)

    def __call__(self, word, **criteria):
        traversals = tuple(self.analyzer.analyze(word))
        criteria_stack = list(criteria.items())
        criteria_stack.reverse()
        while criteria_stack:
            if len(traversals) == 1:
                (traversal,) = traversals
                return traversal.analysis
            attr, attr_vals = criteria_stack.pop()
            filtered = tuple((t for t in traversals if getattr(t, attr) in attr_vals))
            traversals = filtered or traversals
        for traversal in traversals:
            return traversal.analysis


def lemmatizer(*args, automaton_type="lemma", **kwargs):
    return Lemmatizer(automata(*args, **kwargs), automaton_type)
