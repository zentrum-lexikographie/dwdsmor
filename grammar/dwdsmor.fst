% dwdsmor.fst
% Version 6.1
% Andreas Nolda 2023-03-21

#include "symbols.fst"
#include "num.fst"
#include "stemtype.fst"
#include "wf.fst"
#include "infl.fst"
#include "markers.fst"
#include "phon.fst"
#include "punct.fst"
#include "cap.fst"
#include "cleanup.fst"


% lexicon

$LEX$ = "dwds.lex"


% numbers

$LEX$ = $LEX$ | $NUM$


% levels

% use surface level of derivation stems and compounding stems
% also for their analysis level
$LEX$ = ( $LEX$ || $BaseStemFilter$) | \
        (^$LEX$ || $DerStemFilter$)  | \
        (^$LEX$ || $CompStemFilter$)


% cleanup of level-specific symbols

$LEX$ = $CleanupInflAnalysis$ || $LEX$

$LEX$ = $LEX$ || $CleanupIndex$


% surface triggers

$LEX$ = $LEX$ || $SurfaceTriggers$


% stem types

$BaseStems$ = $LEX$ || $BaseStemFilter$
$DerStems$  = $LEX$ || $DerStemFilter$
$CompStems$ = $LEX$ || $CompStemFilter$

$BaseStemsDC$ = $StemDCAnalysis$ || $BaseStems$ || $StemDC$
$DerStemsDC$  = $StemDCAnalysis$ || $DerStems$  || $StemDC$
$CompStemsDC$ = $StemDCAnalysis$ || $CompStems$ || $StemDC$


% word formation

% morpheme-boundary triggers

$CompBreakNone$ = <CB>:<^none>
$CompBreakHyph$ = {<HB>\-<CB>}:{<^hyph>}

$DerBreak$ =  <DB>:<>

% affixes

$Pref-un$   = <Prefix> un $DerBreak$
$PrefUC-un$ = <Prefix> Un $DerBreak$

$Suff-chen$ = <DB>:<Suffix> chen <NN> <base> <native> <>:<NNeut_s_x>
$Suff-lein$ = <DB>:<Suffix> lein <NN> <base> <native> <>:<NNeut_s_x>

% derived base stems

$DerStemsDim$  = $DerStems$ || $DerStemDimFilter$

$DerBaseStems$ = $Pref-un$   $BaseStems$   | \
                 $PrefUC-un$ $BaseStemsDC$ | \
                 $DerStemsDim$ $Suff-chen$ | \
                 $DerStemsDim$ $Suff-lein$ || $DerFilter$

$BaseStems$ = $BaseStems$ | $DerBaseStems$

$DerBaseStemsDC$ = $StemDCAnalysis$ || $DerBaseStems$ || $StemDC$

$BaseStemsDC$ = $BaseStemsDC$ | $DerBaseStemsDC$

$BASE$ = $BaseStems$

% derived compounding stems

$DerCompStems$ = $Pref-un$ $CompStems$ | $PrefUC-un$ $CompStemsDC$ || $DerFilter$

$CompStems$ = $CompStems$ | $DerCompStems$

$DerCompStemsDC$ = $StemDCAnalysis$ || $DerCompStems$ || $StemDC$

$CompStemsDC$ = $CompStemsDC$ | $DerCompStemsDC$

% compounds

$COMP$ = $CompStems$ \
         ($CompBreakHyph$ $CompStems$ | $CompBreakNone$ $CompStemsDC$)* \
         ($CompBreakHyph$ $BaseStems$ | $CompBreakNone$ $BaseStemsDC$) || $CompFilter$

$LEX$ = $BASE$ | $COMP$


% cleanup of word-formation-related symbols

$LEX$ = $CleanupWFAnalysis$ || $LEX$

$LEX$ = $LEX$ || $CleanupWF$


% inflection

$MORPH$ = $LEX$ $INFL$ || $InflFilter$


% inflection markers

$MORPH$ = $MORPH$ || $MarkerGe$

$MORPH$ = $MORPH$ || $MarkerZu$

$MORPH$ = $MORPH$ || $MarkerImp$


% morpheme-boundary triggers generated from entry types

$MORPH$ = $MORPH$ || $BoundaryTriggers$


% (morpho)phonology

$MORPH$ = <>:<WB> $MORPH$ <>:<WB> || $PHON$

$MORPHAnalysis$ = (_$MORPH$) || $PHONAnalysis$

$MORPH$ = (^_$MORPHAnalysis$) || $MORPH$


% punctuation

$MORPH$ = $MORPH$ | $PUNCT$


% morpheme-boundary markers

$MORPH$ = $MarkerBoundaryAnalysis$ || $MORPH$


% cleanup of orthography-related symbols

$MORPH$ = $CleanupOrthAnalysis$ || $MORPH$

% cleanup of lemma and paradigm indices

$MORPH$ = $CleanupIndexAnalysis$ || $MORPH$

% cleanup of morpheme-boundary triggers

$MORPH$ = $MORPH$ || $CleanupBoundary$


% capitalisation

$MORPH$ = $MORPH$ | <CAP>:<> ($MORPH$ || $CAP$)


% the resulting automaton

$MORPH$
