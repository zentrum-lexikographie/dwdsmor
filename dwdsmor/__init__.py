from .automaton import automata, lemmatizer
from .version import __version__

__all__ = ["automata", "lemmatizer", "__version__"]
