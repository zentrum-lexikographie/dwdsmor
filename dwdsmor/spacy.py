from typing import Iterable

from spacy.language import Language
from spacy.tokens.token import Token

import dwdsmor.tag.hdt as hdt

from . import lemmatizer
from .automaton import Lemmatizer

Token.set_extension("dwdsmor", default=None)


def morph(token_morph, k):
    v = ",".join(token_morph.get(k))
    return v if v else None


def lemmatize_token(lemmatizer: Lemmatizer, token: Token):
    token_morph = token.morph
    token_criteria = hdt.criteria(
        token.tag_,
        morph(token_morph, "Number"),
        morph(token_morph, "Gender"),
        morph(token_morph, "Case"),
        morph(token_morph, "Person"),
        morph(token_morph, "Tense"),
        morph(token_morph, "Degree"),
        morph(token_morph, "Mood"),
        morph(token_morph, "VerbForm"),
    )
    token._.dwdsmor = lemmatizer(token.text, **token_criteria)
    return token


def lemmatize(lemmatizer: Lemmatizer, tokens: Iterable[Token]):
    for token in tokens:
        lemmatize_token(lemmatizer, token)
    return tokens


class Component:
    def __init__(self, automata_location=None):
        self.lemmatizer = lemmatizer(automata_location)

    def __call__(self, doc):
        lemmatize(self.lemmatizer, doc)
        return doc


@Language.factory("dwdsmor", default_config={"automata_location": None})
def create_component(nlp: Language, name: str, automata_location: str | None):
    return Component(automata_location)
