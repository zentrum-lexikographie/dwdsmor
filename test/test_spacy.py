import spacy
from datasets import load_dataset
from pytest import fixture

import dwdsmor
import dwdsmor.spacy


@fixture(scope="module")
def nlp():
    nlp = spacy.load("de_hdt_lg")
    nlp.add_pipe("dwdsmor")
    return nlp


@fixture(scope="module")
def lemmatizer():
    return dwdsmor.lemmatizer()


@fixture(scope="module")
def sentences():
    ds = load_dataset(
        "universal_dependencies",
        "de_gsd",
        split="train",
        trust_remote_code=True,
    )
    return tuple(s["text"] for s in ds.select(range(100)))


def test_lemmatisation(nlp, lemmatizer, sentences, snapshot):
    sentences = sentences[:10]
    docs = nlp.pipe(sentences)
    tokens = ((t.text, t.tag_, t.lemma_, t._.dwdsmor_lemma) for d in docs for t in d)
    assert tuple(tokens) == snapshot
