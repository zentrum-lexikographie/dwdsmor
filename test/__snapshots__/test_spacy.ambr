# serializer version: 1
# name: test_lemmatisation
  tuple(
    tuple(
      'Sehr',
      'ADV',
      'Sehr',
      'sehr',
    ),
    tuple(
      'der',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'Kundenservice',
      'NN',
      'Service',
      'Kundenservice',
    ),
    tuple(
      'Die',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'im',
      'APPR_ART',
      'in der',
      'in',
    ),
    tuple(
      'Nette',
      'ADJA',
      'Nette',
      'nett',
    ),
    tuple(
      'zur',
      'APPR_ART',
      'zu der',
      'zu',
    ),
    tuple(
      'der',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'im',
      'APPR_ART',
      'in der',
      'in',
    ),
    tuple(
      'Vital',
      'NN',
      'Vital',
      'vital',
    ),
    tuple(
      'Positives',
      'NN',
      'Positiv',
      'positiv',
    ),
    tuple(
      'Ob',
      'KOUS',
      'Ob',
      'ob',
    ),
    tuple(
      'der',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'den',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'Behandlungsräumen',
      'NN',
      'Behandlungsräumen',
      'Behandlungsraum',
    ),
    tuple(
      'den',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'zugeschnittenen',
      'ADJA',
      'zuschneiden',
      'zugeschnitten',
    ),
    tuple(
      'äußerst',
      'ADV',
      'äußerst',
      'äußern',
    ),
    tuple(
      'das',
      'PDS',
      'der',
      'die',
    ),
    tuple(
      'zum',
      'APPR_ART',
      'zu der',
      'zu',
    ),
    tuple(
      'der',
      'PRELS',
      'der',
      'die',
    ),
    tuple(
      'anstrengend',
      'ADJD',
      'anstrengen',
      'anstrengend',
    ),
    tuple(
      'dieser',
      'PDAT',
      'dieser',
      'diese',
    ),
    tuple(
      'jede',
      'PIDAT',
      'jeder',
      'jede',
    ),
    tuple(
      'Das',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'nächste',
      'ADJA',
      'NULL',
      'nah',
    ),
    tuple(
      'einen',
      'ART',
      'ein',
      'eine',
    ),
    tuple(
      'der',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'Eine',
      'ART',
      'ein',
      'eine',
    ),
    tuple(
      'einen',
      'ART',
      'ein',
      'eine',
    ),
    tuple(
      'der',
      'PRELS',
      'der',
      'die',
    ),
    tuple(
      'Die',
      'ART',
      'der',
      'die',
    ),
    tuple(
      'mehr',
      'ADV',
      'mehr',
      'sehr',
    ),
  )
# ---
